test/catch2/ext/test_key_provider.cpp

```diff
@@ -0,0 +1,318 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "ext/test/key_provider/key_provider.h"
+
+#include "wrappers/connection_wrapper.h"
+#include "utils.h"
+
+#include "wiredtiger.h"
+#include "wt_internal.h"
+
+#include <catch2/catch.hpp>
+
+#include <functional>
+#include <iostream>
+#include <memory>
+
+/*
+ * kp_fixture
+ *     Test fixture for the key provider extension tests.
+ */
+struct kp_fixture {
+    utils::shared_library lib{KEY_PROVIDER_EXTENSION};
+
+    using extension_init_t = decltype(&wiredtiger_extension_init);
+    extension_init_t extension_init = nullptr;
+
+    std::unique_ptr<connection_wrapper> conn;
+
+    KEY_PROVIDER *kp = nullptr;
+    using kp_ptr_t = std::unique_ptr<KEY_PROVIDER, std::function<void(KEY_PROVIDER *)>>;
+
+    ~kp_fixture()
+    {
+        kp_reset();
+    }
+
+    kp_fixture()
+        : extension_init(lib.get<extension_init_t>("wiredtiger_extension_init")),
+          conn(std::make_unique<connection_wrapper>(DB_HOME, "create,in_memory"))
+    {
+    }
+
+    kp_ptr_t
+    kp_init(const char *config)
+    {
+        const char *ext_config[] = {config, nullptr};
+
+        int ret =
+          extension_init(conn->get_wt_connection(), reinterpret_cast<WT_CONFIG_ARG *>(ext_config));
+        REQUIRE(ret == 0);
+
+        REQUIRE(kp == nullptr);
+        kp = reinterpret_cast<KEY_PROVIDER *>(conn->get_wt_connection_impl()->key_provider);
+        REQUIRE(kp != nullptr);
+
+        return kp_ptr_t(kp, [this](KEY_PROVIDER *k) {
+            assert(k == this->kp);
+            this->kp_reset();
+        });
+    }
+
+    void
+    kp_reset()
+    {
+        if (kp != nullptr) {
+            int ret = kp->iface.terminate(&kp->iface, session());
+            if (ret != 0) {
+                WARN("Error terminating key provider: " << ret);
+            }
+            kp = nullptr;
+        }
+        conn->get_wt_connection_impl()->key_provider = nullptr;
+    }
+
+    WT_SESSION *
+    session()
+    {
+        return &conn->get_wt_connection_impl()->default_session->iface;
```

Comment:
I wouldn't recommend using the default session. The default session is only meant for internal WT use, it may lead to undefined behaviour. I would create your own one like this:
```
    REQUIRE(conn->open_session(conn, NULL, NULL, &session) == 0);
```
reference: test_sub_level_error_rollback.cpp